cpp/arcticdb/vector_db/vector_db.hpp

```diff
@@ -103,14 +103,55 @@ namespace arcticdb {
 
         TopKClause(std::vector<double> query_vector, uint8_t k);
 
-        [[nodiscard]] Composite<ProcessingSegment> process(
+        [[nodiscard]] Composite<ProcessingUnit> process(
                 std::shared_ptr<Store> store,
-                Composite<ProcessingSegment> &&p
+                Composite<ProcessingUnit> &&p
         ) const;
 
-        [[nodiscard]] std::optional<std::vector<Composite<ProcessingSegment>>> repartition(
-                ARCTICDB_UNUSED std::vector<Composite<ProcessingSegment>> &&) const {
-            return std::nullopt;
+        [[nodiscard]] std::vector<Composite<SliceAndKey>> structure_for_processing(
+                std::vector<SliceAndKey>& slice_and_keys, ARCTICDB_UNUSED size_t) const {
+            return structure_by_column_slice(slice_and_keys);
+        }
+
+        [[nodiscard]] std::optional<std::vector<Composite<ProcessingUnit>>> repartition(
+                std::vector<Composite<ProcessingUnit>> &&c,
+                const std::shared_ptr<Store>& store) const {
+            auto comps = std::move(c);
+            TopK top_k(k_);
+            for (auto &comp : comps) {
+                comp.broadcast([&store, &top_k](auto &proc) {
+                    for (const auto& slice_and_key: proc.data()) {
+                        const std::vector<std::shared_ptr<Column>>& columns = slice_and_key.segment(store).columns();
+                        for (auto&& [idx, col]: folly::enumerate(columns)) {
+                            col->type().visit_tag([&top_k, &col=col, &slice_and_key, &store, idx=idx](auto type_desc_tag) {
+                                using TDT = decltype(type_desc_tag);
+                                using raw_type = typename TDT::DataTypeTag::raw_type;
+
+                                if constexpr (is_floating_point_type(TDT::DataTypeTag::data_type)) {
+                                    top_k.try_insert(
+                                            col->template scalar_at<raw_type>(col->last_row()).value(),
```

Comment:
I would add a comment somewhere about the structure of the input `std::vector<Composite<ProcessingUnit>>` and the columns in them to explain why this works.

Reply:
Well that the input `ProcessingUnit`s represent the top_k of subsets of the columns, and that the distance is stored in the last row. That isn't obvious at all from just this method.